app/validators/uri_validator.rb

```diff
@@ -1,53 +1,102 @@
 # frozen_string_literal: true
 
 class UriValidator < ActiveModel::EachValidator
-  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
-  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
-  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
-
-  def validate_each(record, attribute, value)
-    valid = begin
-              uri = URI.parse(value)
-              valid_scheme?(uri.scheme) && valid_host?(uri.host) && !forbidden_part?(record, uri)
-            rescue URI::InvalidURIError
-              false
-            end
-    record.errors.add(attribute, options[:message] || :invalid) unless valid
-    valid
+  def validate_each(record, attribute_name, attribute_value)
+    errors = UriPartsValidator.validate(attribute_value, self.class.options_for_record(options, record))
+    errors.each { |error| record.errors.add(attribute_name, error) }
+    errors.empty?
   end
 
-  def valid_scheme?(scheme)
-    accepted_scheme = options.fetch(:scheme, DEFAULT_ACCEPTED_SCHEMES)
-    case accepted_scheme
-    when Regexp
-      scheme =~ accepted_scheme
-    else
-      [*accepted_scheme].include? scheme
+  # This smells :reek:ManualDispatch
+  def self.options_for_record(options, record)
+    options.each_with_object({}) do |(options_name, options_value), opts|
+      opts[options_name] = options_value.respond_to?(:call) ? options_value.call(record) : options_value
     end
   end
 
-  def valid_host?(host)
-    return false if host.blank?
-    (host.size <= 255) && valid_host_labels?(host)
-  end
+  class UriPartsValidator
```

Comment:
I was thinking... so now `UriValidator` is the actual `ActiveModel::Validator`. It uses this class that does all the hard work – e.g. parsing the value into "parts" of an URI, actually validating each part according to the options and the RFC. Then `UriValidator` only adds the error message. Do we really need two classes? Is it all "URI validation"? 🤔

Reply:
> Do we really need two classes? 

I do not think that we ever really **need** more than one single file for the entire program to work, but to maintain it is another story... same here, **need** is a too strong word and I didn't add it out of necessity but simply because I believe that makes our lives easier.
That class is not mutable and does not have side effects either. It is just one input that implies one output, always, idempotently, no matter how many times it is called. At the same time, not so many parameters being passed everywhere, but just relying on its instance variables, which aren't many.

> Is it all "URI validation"?

I think that this is the problem and what actually makes it confusing... that I didn't write a proper name and it seems like both are "URI validation". I definitely have to fix that :+1: